pkg/upstream/upstream.go

```diff
@@ -198,3 +220,42 @@ func (up *Upstream) IsNormal() bool {
 func (up *Upstream) IsClosed() bool {
 	return atomic.LoadInt32(&up.status) == closed
 }
+
+func (up *Upstream) unhold() {
+	up.hcMu.mu.Lock()
+	defer up.hcMu.mu.Unlock()
+	up.hcMu.hc--
+	if up.hcMu.hc < 0 {
+		log.Panic("upstream's hc should never less than 0", zap.Uint64("upstreamID", up.ID))
+	}
+	if up.hcMu.hc == 0 && up.hcMu.idleTime.IsZero() {
+		up.hcMu.idleTime = time.Now()
+	}
+}
+
+func (up *Upstream) hold() {
```

Comment:
The names of `hold` and `hc` are somewhat confusing, perhaps we can find a more appropriate name?